cpp/src/arrow/array/builder_nested.h

```diff
@@ -240,20 +240,38 @@ class ARROW_EXPORT MapBuilder : public ArrayBuilder {
   /// \brief Start a new variable-length map slot
   ///
   /// This function should be called before beginning to append elements to the
-  /// key and value builders
+  /// key and item builders
   Status Append();
 
   Status AppendNull() final;
 
   Status AppendNulls(int64_t length) final;
 
+  /// \brief Get builder to append keys.
+  ///
+  /// Append a key with this builder should be followed by appending
+  /// an item or null value with item_builder().
   ArrayBuilder* key_builder() const { return key_builder_.get(); }
+
+  /// \brief Get builder to append items
+  ///
+  /// Appending an item with this builder should have been proceeded
+  /// by appending a key with key_builder().
   ArrayBuilder* item_builder() const { return item_builder_.get(); }
 
+  /// \brief Get builder to add Map entries as struct values.
+  ///
+  /// This is used instead of key_builder()/item_builder() and allows
+  /// the Map to be built as a list of struct values.
+  ArrayBuilder* value_builder() const { return list_builder_->value_builder(); }
+
   std::shared_ptr<DataType> type() const override {
     return map(key_builder_->type(), item_builder_->type(), keys_sorted_);
   }
 
+ protected:
+  Status AdjustStructBuilderLength();
```

Comment:
You could mark this "inline" to encourage the compiler to fold it into its callers.